javascript/examples/rollup/public/in-service-worker/service-worker.js

```diff
@@ -0,0 +1,87 @@
+// Firefox is not heavy on forwarding logs from Service Workers
+// to the main window, so instead, we'll use a BroadcastChannel
+// to help debugging. A bit blunt, but that'll do
+const channel = new BroadcastChannel("service-worker-log")
+const originalLog = console.log
+console.log = function (...args) {
+  try {
+    channel.postMessage(args)
+  } catch (error) {
+    // There'll be one point where I'll try to log something
+    // that's not serialisable for `postMessage` so that'll
+    // avoid crashing.
+    originalLog(...args)
+  }
+}
+
+// Import the bundled script for automerge, which will add an `AutomergePromise`
+// allowing to wait for the automerge API once all the event listeners
+// of the service worker have been set up synchronously
+self.importScripts("/in-service-worker/automerge.js")
+
+// Little trickery to have a way to resolve the promise only late
+// With that pattern, we get a reference we can pass to `waitUntil`
+// or use as a way to delay responses to `isReady`, while delaying
+// when the Promise is resolved till the Automerge API has been loaded
+// Thankfully the function passed to the Promise constructor
+// runs synchronously, allowing to register the listeners OK
+const ready = new Promise(markReady => {
+  self.addEventListener("install", event => {
+    console.log("Installing service worker", new Date().toISOString())
+    self.skipWaiting()
+
+    event.waitUntil(ready)
+  })
+
+  self.addEventListener("activate", () => {
+    console.log("Service worker activated")
+  })
+
+  self.addEventListener("message", event => {
+    console.log("Message from page", event)
+    const handler = MESSAGE_HANDLERS[event.data.action]
+    if (handler) {
+      handler(event)
+    }
+  })
+
+  const MESSAGE_HANDLERS = {
+    async isReady(event) {
+      console.log("Checking if server is ready", ready)
+      await ready
+      event.source.postMessage({ action: "ready" })
```

Comment:
Something that I came across since I raised the PR. Not sure if you'd prefer it as an opt-in rather than a default, let me know 😊 

```suggestion
      // Make the service worker start handling fetch events from the page straight away
      // which can be necessary when users reload `Ctrl/Command-Shift-R`, 
      // which bypasses service worker
      // See: https://web.dev/articles/service-worker-lifecycle#shift-reload
      await self.clients.claim()
      event.source.postMessage({ action: "ready" })
```

Reply:
No, if you need this you need it. This is good but @romaricpascal I don't see you handling controllerchange here.

Reply:
🤔 I'll need to have some further thoughts on whether it's a need in all situations. Same would go for the `self.skipWaiting()` on line 31. 

I'm wondering if these two statements are not app dependent, unlike the rest of the code which is really tied to automerge needing to load asynchronously and coordinating the page with that given `navigator.serviceWorker.ready` wouldn't provide the info that automerge has actually loaded.

Depending on your app, you may or may not want to activate a new worker at a custom time rather than straight after install (eg. coodinating with the UI via a `BroadcastChannel` rather than automatically), and may or may not want to claim clients straight away (though this seems more necessary for shift-reloads and first run), which in turn may or may not lead you to want to use `controllerchange` (which only matters if the JS on the page and the one in the worker are different versions of the app). This seems to fall a bit in the same area as the `update()` method on registration and `updatefound` events, where apps will have different use depending on the feature they provide.

Reply:
@pvh I've had a bit more thinking on what's the minimum necessary and that's where I got to:
1. I'd say the `clients.claim()` is necessary. In the context of running Automerge in a Service Worker, you'll likely want to be sure that your page is controlled by the service worker, which means you want the worker to take control on first run (or Shift-reloads).
2. The `self.skipWaiting` however, is a mater of how you want a new version of the worker to take over, so I'd be keen to leave it out of the example as it'll depend on how the app wants to handle its service worker updates (which will vary from one to another).
3. Similarly `controllerchange` only occurs when a new controller is replacing an existing one if I understood correctly, so again a matter of how the app wants to handle its service worker updates.
4. And `update` and its related `updatefound` event also depend on how the app plans to handle update.

In the end, I'm thinking of:
- removing the `self.skipWaiting` bit from the example
- adding a note in the service worker section of the readme explaining that:
    - this example only deals with bundling automerge for use in a service worker and the coordination with loading the WASM
    - how the service worker updates will be dependent from one application to another, and you can find a trove of information on the different mechanisms at your disposal in [The service worker lifecycle article on web.dev](https://web.dev/articles/service-worker-lifecycle).

Does that sounds a good way forward?